openpgp/key_generation.go

```diff
@@ -120,6 +120,91 @@ func NewEntity(name, comment, email string, config *packet.Config) (*Entity, err
 	}, nil
 }
 
+// AddSigningSubkey adds a signing keypair as a subkey to the Entity.
+// If config is nil, sensible defaults will be used.
+func (e *Entity) AddSigningSubkey(config *packet.Config) error {
+	creationTime := config.Now()
+
+	subPrivRaw, err := newSigner(config)
+	if err != nil {
+		return err
+	}
+	sub := packet.NewSignerPrivateKey(creationTime, subPrivRaw)
+
+	subkey := Subkey{
+		PublicKey:  &sub.PublicKey,
+		PrivateKey: sub,
+		Sig: &packet.Signature{
+			CreationTime:              creationTime,
+			SigType:                   packet.SigTypeSubkeyBinding,
+			PubKeyAlgo:                e.PrimaryKey.PubKeyAlgo,
+			Hash:                      config.Hash(),
+			FlagsValid:                true,
+			FlagSign:                  true,
+			IssuerKeyId:               &e.PrimaryKey.KeyId,
+		},
+	}
+
+	embeddedSig := &packet.Signature{
+		CreationTime: creationTime,
+		SigType:      packet.SigTypePrimaryKeyBinding,
+		PubKeyAlgo:   subkey.PublicKey.PubKeyAlgo,
+		Hash:         config.Hash(),
+		IssuerKeyId:  &e.PrimaryKey.KeyId,
+	}
+	err = embeddedSig.CrossSignKey(subkey.PublicKey, e.PrimaryKey, subkey.PrivateKey, config)
+	if err != nil {
+		return err
+	}
+	subkey.Sig.EmbeddedSignature = embeddedSig
+
+	subkey.PublicKey.IsSubkey = true
+	subkey.PrivateKey.IsSubkey = true
```

Comment:
In the end, `embeddedSig.PubkeyAlgo` depends on `subkey` and `subkey` depends on `embeddedSig`. They are just values but looks circular. 

Maybe create first `embeddedSig` (and use e.g. `config.PublicKeyAlgorithm()`) and then simply set 
```
EmbeddedSignature: &embeddedSig
```
or something similar when defining `subkey`?

Reply:
Yeah, good point. I've inlined the embedded signature creation into the subkey creation, to still keep the packet order but fix the circularity.